src/core/cryptor.rs

```diff
@@ -0,0 +1,57 @@
+//! Cryptor module
+//!
+//! This module contains the [`Cryptor`] trait which is used to implement
+//! encryption and decryption of published data.
+
+use crate::core::error::PubNubError;
+
+/// This trait is used to encrypt and decrypt messages sent to the
+/// [`PubNub API`].
+///
+/// It is used by the [`dx`] modules to encrypt messages sent to PubNub and
+/// returned by the [`PubNub API`].
+///
+/// To implement this trait, you must provide `encrypt` and `decrypt` methods
+/// that takes a `&[u8]` and returns a `Result<Vec<u8>, PubNubError>`.
+///
+/// You can implement this trait for your own types, or use one of the provided
+/// features to use a crypto library.
+///
+/// # Examples
+/// ```
+/// use pubnub::core::{Cryptor, error::PubNubError};
+///
+/// struct MyCryptor;
+///
+/// impl Cryptor for MyCryptor {
+///     fn encrypt<'en, T: Into<&'en [u8]>>(&self, source: T) -> Result<Vec<u8>, PubNubError> {
+///         // Encrypt provided data here
+///
+///         Ok(vec![])
+///     }
+///
+///     fn decrypt<'de, T: Into<&'de [u8]>>(&self, source: T) -> Result<Vec<u8>, PubNubError> {
+///         // Decrypt provided data here
+///
+///         Ok(vec![])
+///     }
+/// }
+/// ```
+///
+/// [`dx`]: ../dx/index.html
+/// [`PubNub API`]: https://www.pubnub.com/docs
+pub trait Cryptor {
+    /// Decrypt provided data.
+    ///
+    /// # Errors
+    /// Should return an [`PubNubError::EncryptionError`] if provided data can't
+    /// be encrypted or underlying cryptor misconfigured.
+    fn encrypt<'en, T: Into<&'en [u8]>>(&self, source: T) -> Result<Vec<u8>, PubNubError>;
```

Comment:
@kleewho @Xavrax is there a way to check, at run-time, whether a provided implementation has either `encrypt` or `decrypt` methods? I thought it is impossible. 
Other option which I see, if we want them separate to ask the user to provide separately implemented encryptor and / or decryptor during client configuration.

Reply:
You can require that type that you are providing must have one or both traits e.g.:
```rust
fn foo<T>(_:T) where T: Encrypt
fn bar<T>(_:T) where T: Decrypt
fn baz<T>(_:T) where T: Encrypt + Decrypt
```

Reply:
@Xavrax ok, so we do need separate options in our builder in this case, because one thing is user provide them to the PubNub client and the other thing client internally know what exactly can be done with provider crypto provider. If they will be stored in different variables - OK, but overkill as for me.

Reply:
I'm not saying that we have to do it. Or that we have to do it now. I'm fine for now as it is

Reply:
@kleewho spoke with @Xavrax, and it looks like a bit overkill, at least for now.